packages/core/src/config/types.ts

```diff
@@ -37,16 +43,24 @@ export type PreprocessorConfig =
 
 export type DecoratorConfig = PreprocessorConfig;
 
-export type StyleguideRawConfig = {
+export type StyleguideRawConfig<T = undefined> = {
   plugins?: (string | Plugin)[];
   extends?: string[];
   doNotResolveExamples?: boolean;
   recommendedFallback?: boolean;
 
-  rules?: Record<string, RuleConfig>;
-  oas2Rules?: Record<string, RuleConfig>;
-  oas3_0Rules?: Record<string, RuleConfig>;
-  oas3_1Rules?: Record<string, RuleConfig>;
+  rules?: T extends 'built-in'
+    ? { [key in BuiltInCommonRuleId | BuiltInCommonOASRuleId]: RuleConfig }
```

Comment:
This is how we can enforce our built-in rulesets to be in sync with the rules that we validate in [redocly-yaml.ts](https://github.com/Redocly/redocly-cli/pull/1308/files#diff-2528117ea3cbe25dfbfd0583c9b11ba7baa7ef347184dfba32a73ec486aa1a5eR7).

Reply:
Maybe we can add a separate type for this, something like

```ts
type RuleMap<T, V> = T extends 'built-in' ? { [key in V]: RuleConfig } : Record<string, RuleConfig>;
```

and use it 

```ts
  rules?: RuleMap<T, BuiltInCommonRuleId | BuiltInCommonOASRuleId>;
  oas2Rules?: RuleMap<T, BuiltInOAS2RuleId>;
  oas3_0Rules?: RuleMap<T, BuiltInOAS3RuleId>;
  oas3_1Rules?: RuleMap<T, BuiltInOAS3RuleId>;
```